devices/MulticastDns/MulticastDnsService.cs

```diff
@@ -0,0 +1,139 @@
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MulticastDnsService" /> class.
+        /// </summary>
+        public MulticastDnsService()
+        {
+            _multicastAddress = IPAddress.Parse(MulticastDnsAddress);
+            _client = new(new IPEndPoint(IPAddress.Any, MulticastDnsPort));
+        }
+
+        /// <summary>
+        /// Start the worker thread that will listen for Multicast packets.
+        /// </summary>
+        public void Start()
+        {
+            if (!_listening)
+            {
+                _listening = true;
+                new Thread(Run).Start();
+            }
+        }
+
+        /// <summary>
+        /// Stop the worker thread that is listening for Multicast packets.
+        /// </summary>
+        public void Stop() => _listening = false;
+
+        /// <summary>
+        /// The delegate that will be invoked when a Multicast DNS message is received.
+        /// </summary>
+        /// <param name="sender">The MulticastDNSService instance that received the message.</param>
+        /// <param name="e">The MessageReceivedEventArgs containing the received message.</param>
+        public delegate void MessageReceivedEventHandler(object sender, MessageReceivedEventArgs e);
+
+        /// <summary>
+        /// The event that is raised when a Multicast DNS message is received.
+        /// </summary>
+        public event MessageReceivedEventHandler MessageReceived;
+
+        /// <summary>
+        /// The delegate that will be invoked when the status of a Multicast DNS service is changed.
+        /// </summary>
+        /// <param name="sender">The MulticastDNSService instance that is reporting its status.</param>
+        /// <param name="e">The MulticastDnsStatusEventArgs containing the status and an optional message.</param>
+        public delegate void MulticastDnsStatusChangedEventHandler(object sender, MulticastDnsStatusEventArgs e);
+
+        /// <summary>
+        /// The event that is raised when the status of a Multicast DNS service is changed.
+        /// </summary>
+        public event MulticastDnsStatusChangedEventHandler StatusChanged;
+
+        /// <summary>
+        /// Sends a Multicast DNS message.
+        /// </summary>
+        /// <param name="message">The message to be sent.</param>
+        public void Send(Message message) => _client.Send(message.GetBytes(), new(_multicastAddress, MulticastDnsPort));
+
+        private void Run()
+        {
+            try
+            {
+                _client.JoinMulticastGroup(_multicastAddress);
+
+                IPEndPoint multicastEndpoint = new(_multicastAddress, MulticastDnsPort);
+                IPEndPoint remoteEndpoint = new(IPAddress.Any, 0);
+
+                byte[] buffer = new byte[2048];
+
+                while (_listening)
+                {
+                    StatusChanged?.Invoke(this, new MulticastDnsStatusEventArgs(MulticastDnsStatus.Running));
```

Comment:
you should move this one before the while loop as if it's running, it's just running :-)

Reply:
Doh! Good catch :-)

Reply:
That's the principle of having PR review in depth :-) And it takes a lot of time to do it properly!